src/modm/platform/gpio/stm32/connector.hpp.in

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (c) 2017, 2021, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include "base.hpp"
+#include "unused.hpp"
+#include "static_gpio.hpp"
+#include <type_traits>
+#include <bit>
+
+namespace modm::platform
+{
+
+/// @cond
+namespace detail
+{
+
+template< Peripheral peripheral, Gpio::Signal signal, template<Peripheral _> class... Signals >
+struct GpioGetSignal;
+template< Peripheral peripheral, Gpio::Signal signal, template<Peripheral _> class SignalT, template<Peripheral _> class... Signals >
+struct GpioGetSignal<peripheral, signal, SignalT, Signals...>
+{
+	using Gpio = std::conditional_t<
+				(SignalT<peripheral>::Signal == signal),
+				StaticGpio<typename SignalT<peripheral>::Data>,
+				typename GpioGetSignal<peripheral, signal, Signals...>::Gpio
+			>;
+};
+template< Peripheral peripheral, Gpio::Signal signal >
+struct GpioGetSignal<peripheral, signal>
+{
+	using Gpio = GpioUnused;
+};
+
+} // namespace detail
+
+template< Peripheral peripheral, template<Peripheral _> class... Signals >
+struct GpioConnector
+{
+	template< class GpioQuery >
+	static constexpr bool Contains = (
+		std::is_same_v<typename Signals<peripheral>::Data, typename GpioQuery::Data> or ...);
+
+	template< class GpioQuery >
+	static constexpr bool IsValid = not std::is_same_v<typename GpioQuery::Type, GpioUnused>;
+
+	template< Gpio::Signal signal >
+	using GetSignal = typename detail::GpioGetSignal<peripheral, signal, Signals...>::Gpio;
```

Comment:
@chris-durand is there a way to find/filter a type in a parameter pack in *one line*? I tried to google this, but didn't really get anywhere…

Reply:
Do you want to filter for types that satisfy a predicate? I am not aware of any simple one-line solution without using a metaprogramming library.

Reply:
Yeah, then I'll leave it as is.